src/environmentd/tests/server.rs

```diff
@@ -437,6 +436,118 @@ fn test_cancel_dataflow_removal() {
         .unwrap();
 }
 
+// Test that dataflow uninstalls cancelled peeks.
+#[test]
+#[cfg_attr(miri, ignore)] // unsupported operation: can't call foreign function `epoll_wait` on OS `linux`
+fn test_cancel_dataflow_removal() {
+    test_cancellation_cancels_dataflows("SELECT * FROM t AS OF 9223372036854775807");
+}
+
+#[test]
+#[cfg_attr(miri, ignore)] // unsupported operation: can't call foreign function `epoll_wait` on OS `linux`
+fn test_cancel_long_select() {
+    test_cancellation_cancels_dataflows("WITH MUTUALLY RECURSIVE flip(x INTEGER) AS (VALUES(1) EXCEPT ALL SELECT * FROM flip) SELECT * FROM flip;");
+}
+
+#[test]
+#[cfg_attr(miri, ignore)] // unsupported operation: can't call foreign function `epoll_wait` on OS `linux`
+fn test_cancel_insert_select() {
+    test_cancellation_cancels_dataflows("INSERT INTO t WITH MUTUALLY RECURSIVE flip(x INTEGER) AS (VALUES(1) EXCEPT ALL SELECT * FROM flip) SELECT * FROM flip;");
+}
+
+fn test_closing_connection_cancels_dataflows(query: String) {
+    let config = util::Config::default().unsafe_mode();
+    let server = util::start_server(config).unwrap();
+
+    let mut cmd = Command::new("psql");
```

Comment:
This may need to get moved into mzcompose/testdrive to work? That'd require rewriting this in python, but I think that's reasonable. Testdrive already has retry functionality when running testdrive code.